### PR TITLE
Profiler for blocked IpcReceiver::recv()

### LINKED ISSUE #19223: Report time spent in synchronous IPC to the profiler
There are various parts of the code where we send a message to another thread, and as part of that message we include an `IpcSender` (or just a `Sender`) value that allows the other thread to send a response. In general the initiator of the message will immediately call `recv()` on the corresponding receiver, meaning that the thread will block until it receives a message. It would be valuable to report when this happens to the built-in [profiler](https://github.com/servo/servo/wiki/Profiling#generating-timelines) so it's easier to see how often it affects performance. We should build a wrapper around the `Receiver` and `IpcReceiver` types that automatically [profile](https://github.com/servo/servo/blob/d7b4471c8080c0916277b967994bffcade759c3a/components/profile_traits/time.rs#L112) the synchronous receive operation so it's easy to get this information in the future.

Examples of use:
* https://github.com/servo/servo/blob/504ad4de2a426c2bc70161a9190fddaa8728ead2/components/script/dom/screen.rs#L40-L42
* https://github.com/servo/servo/blob/0e62a5829b7c29ae2667a21a439aff1e89201bf3/components/script/dom/storage.rs#L57-L58
* https://github.com/servo/servo/blob/4c538b642e4bdfbf42c522c5a59c258a6d14546e/components/script/dom/webglprogram.rs#L53-L55

In general, cases of `recv()` (but not `try_recv()`) in components/script are suspect.

### PR DESCRIPTION
**Pull request by nakul02**
Implements feature #19223 

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #19223 (github issue number if applicable).

- [x] There are tests for these changes 

WIP.
@jdm - this is the new profiler : "Blocked at IPC Receive"
Should I dig through all the calls to `ipc::channel` and replace them with this profiled `IpcReceiver`?

screenshot from 2018-02-27 01-35-37

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/20132)

### CODE REVIEW

**Comment by jdm** [comment]
Yes, replacing all the users would be really useful!

**Comment by nakul02** [comment]
@jdm - I've gone ahead and replaced calls to `ipc_channel::ipc::channel` to the newly introduced `profile_traits::receiver::channel` where calls were made to `recv` in the `components/script/dom` folder.
Do you want this done project wide for every call to `ipc_channel::ipc::channel`.
In most other places, there may not be a simple way to get a handle on the `ProfilerChan` instance.

**Comment by nakul02** [comment]
Ping @jdm

**Comment by jdm** [comment]
From my point of view, the more we use this API, the more confident we can be that we're actually noticing the biggest offenders. That being said, if you would prefer to merge the changes that you've made so far, we can also continue changing other channels in followup PRs.

**Comment by nakul02** [comment]
@jdm - if you have suggestions or guidelines for components that I should check for, I can go in and change those. Otherwise, for now, I'd prefer to merge this.
Does the name `profiler_traits::receiver` look ok? It is invoked everywhere as `receiver::channel`.

**Comment by jdm** [comment]
@nakul02 What if we used `profiler_traits::ipc` instead? I think this is a good set of channels to start with, and we can file follow up issues for other instances that make sense, because there are a lot of uses of channels where they are expected to spend a lot of time waiting in recv() and we don't want to report that.

**Comment by jdm** [comment]
Also, I bet we could write a test for this functionality in https://github.com/servo/servo/blob/master/tests/unit/profile/time.rs. Something like:
1. create a regular channel to mimic the time profiler
1. create a profiled channel
1. create a thread that waits for 2 seconds then sends a value with the sender
1. call recv() in the original thread
1. call recv() on the time profiler receiver, assert the its reports a realistic value for the time spent waiting on the first recv()
Does that make sense?

**Comment by nakul02** [comment]
@jdm - I've added a test, but can't figure out why its failing for the life of me. This is the output from its failure:
```
---- time::channel_profiler_test stdout ----
	thread 'time::channel_profiler_test' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Os { code: 104, kind: ConnectionReset, message: "Connection reset by peer" })', libcore/result.rs:916:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:68
             at libstd/sys_common/backtrace.rs:57
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:380
   3: std::panicking::default_hook
             at libstd/panicking.rs:390
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:576
   5: std::panicking::begin_panic
             at libstd/panicking.rs:537
   6: std::panicking::begin_panic_fmt
             at libstd/panicking.rs:521
   7: rust_begin_unwind
             at libstd/panicking.rs:497
   8: core::panicking::panic_fmt
             at libcore/panicking.rs:71
   9: core::result::unwrap_failed
             at /checkout/src/libcore/macros.rs:23
  10: <core::result::Result<T, E>>::unwrap
             at /checkout/src/libcore/result.rs:782
  11: profile_tests::time::channel_profiler_test
             at tests/unit/profile/time.rs:59
  12: <F as alloc::boxed::FnBox<A>>::call_box
             at libtest/lib.rs:1441
             at /checkout/src/libcore/ops/function.rs:223
             at /checkout/src/liballoc/boxed.rs:788
  13: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:102


failures:
    time::channel_profiler_test
```

In this PR, there is a new enum type in the `ProfilerMsg` enum to be able to ask for a vector of recorded times, given information about a `ProfilerBuckets`. I've also `Serialize`-d and `Deserialize`-d the data that is sent through the channel (or so I hope), but none of it seems to be doing the trick. 
Could you please take a look at the code from the *last* commit to see if anything pops out to you. If it doesn't any suggestion about a simpler way of testing this will help.

File: tests/unit/profile/time.rs [hunk 1]
```diff
@@ -37,6 +40,29 @@ fn time_profiler_stats_test() {
...
+    thread::spawn(move || {
+        thread::sleep(Duration::from_secs(2));
+        profiled_sender.send(43).unwrap();
+    });
+
+    let val_profile_receiver = profiled_receiver.recv().unwrap();
+    assert_eq!(val_profile_receiver, 43);
+
+    let (sender, receiver) = ipc::channel().unwrap();
+    thread::spawn(move || {
```
Comment (jdm) [inline tests/unit/profile/time.rs]: No need to create a new thread here.

**Review by jdm — CHANGES_REQUESTED** [review]
These changes look great!

**Comment by jdm** [comment]
Hmm. Unfortunately, this is probably hitting a know bug in ipc-channel that we haven't been able to track down yet. Is there any change if you clone the channel before passing it as an argument?

**Comment by nakul02** [comment]
@jdm - cloning the channel didn't help either. Anything else I can try.
What would be the fallback, in case I can't get around the ipc-channel bug?

**Comment by nakul02** [comment]
@jdm - since rebasing with master, the test is now either blocked or deadlocked.

**Comment by nakul02** [comment]
Ping @jdm

**Comment by jdm** [comment]
Sorry, I spent a few hours looking into what could cause the deadlocking behaviour and got sidetracked by other fires that needed to be put out. It's really curious, because I can't reproduce the problem if I import all of the same code into a different project.

**Comment by jdm** [comment]
If you run `./mach test-unit -p profile -v`, it should show you the binary that's being executed. If you run that binary with `--nocapture`, you should be able to see output from the tests, and adding `eprintln!("...")` to parts of the code may help you figure out where it's becoming blocked.

**Comment by nakul02** [comment]
@jdm - I figured out what the problem was. It was in the change I put in while creation of the `time_profile_chan`. That particular unit test now runs and passes.

**Comment by nakul02** [comment]
Ping @jdm

**Comment by jdm** [comment]
There are many more updates to Cargo.lock than this change requires. Did you run `cargo update -a`? Could you revert the changes to that file and run `./mach build` again instead?

**Review by jdm — APPROVED** [review]
Apart from the Cargo.lock changes, this looks great!

**Comment by nakul02** [comment]
Thanks @jdm!
I deleted the Cargo.lock and ran `mach build -d`. That is what caused all those updates to `Cargo.lock`.

**Comment by nakul02** [comment]
@jdm - is there a way to just cause `Taskcluster` to run again?

**Pull request merged**